### PR TITLE
Implement text wrapping

### PR DESCRIPTION
**Pull request by yayhi**
Adds text wrapping support to `TextElement`:

- If the element is set to a fixed width, text will automatically wrap to the next line.
- If the element is set to `autoWidth`, no wrapping will occur.
- In most cases, the logic will only break to a new line on natural breaks in the text (spaces, tabs and hyphens). However, in the case where a single word is too long to fit on a line by itself, the logic will break the word.
- The existing behaviour of supporting manual line breaks (i.e. via `\n` and `\r`) is preserved.
- Existing behaviour around alignment is also preserved.

I'm still doing some manual testing in the engine itself, but added unit tests to cover the various line break and word break cases. I'll update the PR again once I'm done with manual testing.

### CODE REVIEW

File: tests/framework/components/element/test_textelement.html [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        <script type="text/javascript" src="../../../../src/resources/sprite.js"></script>
+        <script type="text/javascript" src="../../../../src/resources/parser/json-model.js"></script>
+        <script type="text/javascript" src="../../../../src/resources/parser/scene.js"></script>
+        <script type="text/javascript" src="../../../../src/asset/asset.js"></script>
+        <script type="text/javascript" src="../../../../src/asset/asset-variants.js"></script>
+        <script type="text/javascript" src="../../../../src/asset/asset-registry.js"></script>
+        <script type="text/javascript" src="../../../../src/backwards-compatibility.js"></script>
+        <script type="text/javascript" src="../../../../src/scene/immediate.js"></script>
+        <script type="text/javascript" src="../../../../src/scene/lightmapper.js"></script>
+        <script type="text/javascript" src="../../../../src/scene/batching.js"></script>
```
Comment (yayhi) [inline tests/framework/components/element/test_textelement.html]: @daredevildave obviously this isn't good :) I'm basically duplicating `dependencies.txt`. However, I couldn't see a way to test `Elements` or `Entities` without including most of these dependencies. Perhaps we could have a way to transform `dependencies.txt` to a list of `<script>` includes and inject it when loading the tests in a browser?

**Comment by vkalpias** [comment]
One comment from is I think we need a `wrapText` property on the Element component which would probably need to be false by default as this will change the behaviour of existing projects. Then in the Editor we would add a checkbox on Text Elements where we can have it true by default.

**Comment by mcmorry** [comment]
About to apply the wrapping by default, could limit future implementation of auto-sizing of the font.
For example, we could set a fixed width, but not to wrap the text, but to make it fit in that size by reducing the font size.
Also, by setting the height we could enable both wrapping and auto-sizing.

**Comment by vkalpias** [comment]
Also for some reason it looks like not all letters are visible sometimes:

screen shot 2018-03-13 at 9 55 36 am

This should read "This is a sentence with many words"

**Comment by yayhi** [comment]
@vkalpias @daredevildave @mcmorry agree re having a flag, added.

@vkalpias thanks – yes there are some cases I still need to fix when actually running within the engine – please see PR description :) I will update the PR once I'm done with that.

**Comment by yayhi** [comment]
Updated to correctly handle left/center/right alignment:

example-1

And also to handle some special cases around alignment when there is trailing whitespace at the end of a line, or when a hyphenated word is wrapped:

example-2

**Comment by willeastcott** [comment]
Very cool. So @yayhi, do you consider this 'done'/mergeable now?

**Comment by yayhi** [comment]
@willeastcott yes, all good from my side.

**Pull request merged**